### PR TITLE
Warn when device group is not mapped in rootless user namespaces

### LINKED ISSUE #28364: --device does not preserve group ownership in rootless --userns=keep-id
### Issue Description

When using `--device /dev/XXX` with `--userns=keep-id --group-add keep-groups`, device nodes are created with nobody:nobody ownership, ignoring the host's group.

This makes them unreadable in the guest, even thought all permissions align for it to be readable.

### Steps to reproduce the issue

```
podman run -i \
	--rm \
	--userns=keep-id \
	--group-add keep-groups \
	--device /dev/input/ \
	archlinux \
	ls -l /dev/input/event24
```

### Describe the results you received

Output:

```
crw-rw---- 1 nobody nobody 13, 88 Mar 26 00:21 /dev/input/event24
```

### Describe the results you expected

Expected device to have the same permissions as in the host:

```
> ls -l /dev/input/event24
crw-rw----    1 root     plugdev    13,  88 Mar 26 01:21 /dev/input/event24
> id
…,105(plugdev),…
```

### podman info output

```yaml
> podman info
host:
  arch: amd64
  buildahVersion: 1.43.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - dmem
  cgroupManager: cgroupfs
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.13-r0
    path: /usr/bin/conmon
    version: 'conmon version 2.1.13, commit: unknown'
  cpuUtilization:
    idlePercent: 96.43
    systemPercent: 1.08
    userPercent: 2.49
  cpus: 24
  databaseBackend: sqlite
  distribution:
    distribution: alpine
    version: 3.24.0_alpha20260127
  eventLogger: file
  freeLocks: 2043
  hostname: hyperion.whynothugo.nl
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.19.8-0-stable
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 3716919296
  memTotal: 67156721664
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    defaultNetwork: podman
    dns:
      package: aardvark-dns-1.16.0-r0
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.16.0
    package: netavark-1.16.1-r0
    path: /usr/libexec/podman/netavark
    version: netavark 1.16.1
  ociRuntime:
    name: runc
    package: runc-1.4.1-r0
    path: /usr/bin/runc
    version: |-
      runc version 1.4.1
      commit: c67132530367ecd401dbe800bf13b96052212f3f
      spec: 1.3.0
      go: go1.26.1
      libseccomp: 2.6.0
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-2026.01.20-r0
    version: |
      pasta 2026_01_20.386b5f5
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user-1000/podman/podman.sock
  rootlessNetworkCmd: pasta
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.3.3-r0
    version: |-
      slirp4netns version 1.3.3
      commit: 944fa94090e1fd1312232cbc0e6b43585553d824
      libslirp: 4.9.1
      SLIRP_CONFIG_VERSION_MAX: 6
      libseccomp: 2.6.0
  swapFree: 60132790272
  swapTotal: 68719472640
  uptime: 191h 43m 34.00s (Approximately 7.96 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /home/hugo/.config/containers/storage.conf
  containerStore:
    number: 4
    paused: 0
    running: 1
    stopped: 3
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/hugo/.local/share/containers/storage
  graphRootAllocated: 1930587799552
  graphRootUsed: 1871489536000
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 61
  runRoot: /run/user-1000/containers
  transientStore: false
  volumePath: /home/hugo/.local/share/containers/storage/volumes
version:
  APIVersion: 5.8.1
  Built: 1773737838
  BuiltTime: Tue Mar 17 09:57:18 2026
  GitCommit: ""
  GoVersion: go1.26.1
  Os: linux
  OsArch: linux/amd64
  Version: 5.8.1
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

On Alpine Linux Edge.

### Additional information

AFAIK, `--device` recreates nodes via mknod rather than bind-mounting, so host ownership is lost entirely.

--device is unusable for any device that relies on group permissions.

Device in question is a gamepad; access is controlled via group memebership.

### PR DESCRIPTION
**Pull request by bhartigautam156**
Fixes: #28364

#### Problem
When running a rootless container with `--device` `/dev/XXX` with `--userns=keep-id` `--group-add keep-groups` , devices that belong to groups not mapped in the user namespace silently appear as `nobody:nobody` inside the container and are completely inaccessible. There is no error message, no warning and the container just starts and the device doesn't work.
For example:
```bash
podman run -i --rm --userns=keep-id --group-add keep-groups --device /dev/input/ alpine ls -l /dev/input/event0
```
Output:
```bash 
crw-rw----  1 nobody  nobody  ...  event0
```
The root cause is that the device's group GID is not listed in `/etc/subgid`. When the rootless user namespace is created, only GIDs in `/etc/subgid` get mapped so anything else shows up as nobody.

I tried fixing this automatically using idmapped mounts and injecting GIDs directly into the namespace, but both approaches hit kernel-level limitations as you can't map a GID that isn't already in `/etc/subgid` without root privileges. Automatically editing system files like `/etc/subgid` was also ruled out because that's an admin decision, not something Podman should do silently.

#### Solution
Since I can't fix the mapping automatically, I made the failure visible and actionable.Before Podman enters the user namespace, it saves two things to a temporary file:
- The host supplementary groups of the current user
- The real host GIDs of any devices passed via `--device`

This is necessary because once inside the user namespace, GIDs get remapped and we lose track of what the original host GIDs were.Then when the device is being set up for the container, it check:
1. What is the real host GID of this device?
2. Is the current user a member of that group on the host?
3. Is that GID listed in `/etc/subgid`?

Based on this, one of two warnings are showed: If the user is not even a member of the device group:
lets say a group input and the device GID is 994
```bash
Device event0 is owned by group 'input' (GID 994).
You are not a member of this group on the host.
Access will be denied inside the container.
To fix run:
sudo usermod -aG input ubuntu
Then log out and log back in.
```
If the user is a member but the GID is not in `/etc/subgid`:
```bash
Device event0 is owned by group 'input' (GID 994).
You are a member of this group on the host, but GID 994 is not in /etc/subgid.
The device will appear as 'nobody:nobody' inside the container and access will be denied.
To fix this, run:
echo "ubuntu:994:1" | sudo tee -a /etc/subgid && podman system migrate
```
After the user runs the suggested fix and tries again, the device works correctly and no warning is shown.

#### A note on how devices appear inside the container
The ownership shown inside the container can be a little misleading. For example, on the host a device may be owned by:
`root:input` but after the required `/etc/subgid` mapping is added, the same device inside the container may appear as:
`nobody:bin` or, when viewed with numeric IDs `65534:1`
This is expected.
- The owner (`nobody` / UID `65534`) is the kernel's overflow UID because the device owner's UID (`root`) is not mapped into the rootless user namespace.
- The group (`bin` / GID `1`) is **not** the host's group name. It is simply the container's name for container GID `1`. In this case, container GID `1` is correctly mapped to the host's `input` group (GID `994`) through `/etc/subgid`.

In other words, the displayed **group name** inside the container is independent of the host's group name. As long as the numeric GID is mapped correctly, the device permissions work as expected. This PR only detects missing mappings and provides actionable warnings; it does not attempt to preserve host group names inside the container.

#### Files changed
`pkg/rootless/rootless_linux.go` : Before entering the user namespace, saves the user's host groups and the GIDs of all devices passed via `--device` to temporary files. This is the only place where the real host GIDs are available.
`pkg/specgen/generate/config_linux.go` : When adding a device to the container spec, checks whether the device's GID is accessible and emits the appropriate warning if not. 

#### Tests added
`pkg/specgen/generate/config_linux_test.go`:  Unit tests for `isGIDInSubgid()` covering: GID inside range, GID outside range, boundary values (first and last GID in range), GID explicitly added as a single entry, empty subgid file, and wrong username. 
`test/e2e/run_device_test.go` : This runs a rootless container with `/dev/input` and verifies that a warning is emitted mentioning either missing subgid mapping or missing group membership. The test is skipped automatically if not running in rootless mode or if /dev/input` is not available on the host.
- The e2e test depends on `/dev/input/` being present on the host. If there is a better way to test this in CI, I am open to suggestions.

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [ ] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

```release-note
Added a warning when rootless containers cannot access devices due to missing GID mappings in `/etc/subgid`, with exact commands to fix the issue.
```

### CODE REVIEW

File: pkg/rootless/rootless_linux.go [hunk 1]
```diff
@@ -279,6 +347,29 @@ func becomeRootInUserNS(stateDir string) (_ bool, _ int, retErr error) {
...
+			} else {
+				logrus.Debugf("Saved host groups to %s: %s", groupsFile, strings.Join(parts, ","))
+			}
+		}
+	}
+
+	// Save device GIDs from --device flags BEFORE re-exec.
+	// After re-exec device GIDs may be unmapped (appear as 65534).
+	SaveDeviceGIDsFromArgs(stateDir)
+
```
Comment (Luap99) [inline pkg/rootless/rootless_linux.go:372]: that it wrong and will not work, parsing os.Args is incorrect and most commands will never get here as we use the c shortcut code after the first podman command all the time, so you can do this exactly once for the first podman command after boot which is not helpful at all.

**Review by Luap99 — CHANGES_REQUESTED** [review]
That is a lot of code for a very niche use case that is ultimately not going to fix anything either other than a warning. It adds unnecessary state for users who do not need it and ultimately the state is possible incorrect either with the fork exec model. We create the userns once, when you capture the groups there and then store them that means the groups can be out of date later when podman is executed from a different context.

I am also not a fan of suggesting to add the group to subgid,the additional mappings which are then different from subuid will get a whole lot more complex.It will also map the group the same random gid at the end of the range so that alone does not grant the container access either.

Aside from the fact that this does not work properly as written.

The simplest solution to keep your group is --group-add keep-groups with crun (works as long as your container does not call setgroups, i.e. when switching uses in the container). That allows access even when it shows up as nobody.

And please note https://github.com/podman-container-tools/podman/blob/main/LLM_POLICY.md

**Comment by bhartigautam156** [comment]
I agree that suggesting edits to /etc/subgid is not a good approach. But I want to clarify that this was not my first choice either as I have also mentioned in the PR description, I had earlier a different approach idmapped mounts and injecting GIDs directly into the namespace. But both hit kernel-level limitations that cannot be worked around without root privileges, which is why I fell back to the this approach.
Also regarding --group-add keep-groups , I tested this again based on your comment with this time setgroups is not called , but still the device is still inaccessible:
```bash
$ podman run -i --rm --userns=keep-id --group-add keep-groups \
  --network=host \
  --device /dev/input/event0 alpine cat /dev/input/event0
cat: can't open '/dev/input/event0': Permission denied
```
```bash
$ strace -e trace=setgroups -f podman run -i --rm --userns=keep-id \
  --group-add keep-groups --network=host \
  --device /dev/input/event0 alpine cat /dev/input/event0 2>&1 | grep setgroups
(no output)
```
 I am open to suggestions on a better approach as this one is not acceptable

**Comment by Luap99** [comment]
did you use crun? keep-groups atm only works with crun as other runtimes do not have this implemented

**Comment by bhartigautam156** [comment]
> did you use crun? keep-groups atm only works with crun as other runtimes do not have this implemented

Yes , I have used crun
```bash
$ podman info | grep -i crun
    name: crun
    package: crun_1.21-1ubuntu3_amd64
    path: /usr/bin/crun
      crun version 1.21
      rundir: /run/user/1000/crun
```

**Comment by bhartigautam156** [comment]
Hi @Luap99, any update on this? I tested keep-groups with crun and setgroups is never called but the device is still inaccessible. Happy to rework the approach entirely if you can point me in the right direction.

**Comment by giuseppe** [comment]
I agree with @Luap99, we don't need this, too much complexity for something that cannot be addressed by Podman anyway
